ksql-engine/src/main/java/io/confluent/ksql/util/HybridQueryIdGenerator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+
+public class HybridQueryIdGenerator implements QueryIdGenerator {
+  private DefaultQueryIdGenerator defaultQueryIdGenerator;
+  private QueryIdGeneratorUsingOffset queryIdGeneratorUsingOffset;
+  private QueryIdGenerator primaryQueryIdGenerator;
+
+  public HybridQueryIdGenerator() {
+    this(0L);
+  }
+
+  public HybridQueryIdGenerator(final long initialValue) {
+    this.defaultQueryIdGenerator = new DefaultQueryIdGenerator(initialValue);
+    this.queryIdGeneratorUsingOffset = new QueryIdGeneratorUsingOffset();
+    this.primaryQueryIdGenerator = this.defaultQueryIdGenerator;
+
+  }
+
+  public void updateOffset(final long offset) {
```

Comment:
What switches the generator back to the other one for legacy queries? 

We could have a command topic that has a period where legacy and new queries are interleaved as nodes in a cluster are upgraded to the new version of KSQL.  So its important we flip the generation strategy back and forth as we process.

Feels like there is a missing test case to cover this, which would highlight this bug.

We can also decouple the generators from the concept of 'offset'. They don't need to care, we can just call it 'nextId'.

I think the API here should be:

```java
   public void activateNewGenerator(final long nextId) {
      newGenerator.setNextId(nextId)
      activeGenerator = newGenerator;
   }

   public void activateOldGenerator() {
      activeGenerator = oldGenerator;
   }

```

Reply:
Makes sense. Would we want to update the old generator counter to "catch" up to the last offset SpecificQueryIdGeneration was at or the offset of the old command that triggers the activation of the legacy generator?

Reply:
I disagree that we should add this complexity to mitigate a migration concern - IMO it's a sensible behavior to fail interleaved queries that come in during upgrade, especially since they may be error-prone anyway (there's a non-negligible chance that the offset and the sequential numbers will match up)

Reply:
After thinking about this more, I've also switched and think we should probably not allow switching between the legacy and the new one. Once commands are being written to the command topic with the flag, useOffsetAsQueryID, the older versions (5.3) will fail to deserialize the records in the command topic because of the flag
```
Caused by: org.apache.kafka.common.errors.SerializationException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "useOffsetAsQueryID"
```
Once the first new command is pushed to the topic, all nodes are required to upgrade, otherwise they can't execute any new commands.

Reply:
linking other discussion to keep it all in one place: https://github.com/confluentinc/ksql/pull/3354/files#r327707794